File: gn-mobile-ios-1.2.1.4576/mmid/MusicIdObjectiveC/GnAudioVisualizeAdapter.m
//
//  GnAudioVisualizeAdapter.m
//  GN_Music_SDK_iOS
//
//  Copyright (c) 2014 Gracenote. All rights reserved.
//

#import "GnAudioVisualizeAdapter.h"


@import CoreAudio;
@import AudioToolbox;

@interface GnAudioVisualizeAdapter()
@property (weak) id <GnAudioSourceDelegate> audioSource;
@property NSUInteger                         numBitsPerSample;
@property NSUInteger                         numChannels;
@property float lastRMS;
@property float maxRMS;
@property id<GnAudioVisualizerDelegate>      audioVisualizerDelegate;

@end

@implementation GnAudioVisualizeAdapter

-(instancetype) initWithAudioSource:(id<GnAudioSourceDelegate>) audioSource audioVisualizerDelegate:(id<GnAudioVisualizerDelegate> )audioVisualizerDelegate
{
    self = [super init];
    
    if(self)
    {
       self.audioSource = audioSource;
       self.lastRMS = 0;
       self.audioVisualizerDelegate = audioVisualizerDelegate;
    }
    
    return self;
}


-(NSUInteger) sourceInit
{
    if ( !self.audioSource )
    {
        return 1;
    }
    
    NSUInteger retVal = [self.audioSource sourceInit];
    
    // get format information for use later
    
    if ( retVal == 0 )
    {
        self.numBitsPerSample = [self.audioSource sampleSizeInBits];
        self.numChannels = [self.audioSource numberOfChannels];
    }
    
    return retVal;
}


-(NSUInteger) numberOfChannels
{
    return self.numChannels;
}


-(NSUInteger) sampleSizeInBits
{
    return self.numBitsPerSample;
}

-(NSUInteger)  samplesPerSecond
{
    if ( self.audioSource == nil )
    {
        return 0;
    }
    
    return self.audioSource.samplesPerSecond;
}

-(NSUInteger) getData: (unsigned char*)dataBuffer dataSize: (NSUInteger)dataSize
{
    if ( !self.audioSource)
    {
        return 0;
    }
    
    NSUInteger bytesRead = [self.audioSource getData:dataBuffer dataSize:dataSize];
    
    float RMS = [self rms16:(signed char*) dataBuffer bufferSize:bytesRead];
    
    if(self.audioVisualizerDelegate && [self.audioVisualizerDelegate conformsToProtocol:@protocol(GnAudioVisualizerDelegate)] && [self.audioVisualizerDelegate respondsToSelector:@selector(RMSDidUpdateByValue:)])
    {
        [self.audioVisualizerDelegate RMSDidUpdateByValue: RMS/10];
    }

    
    
    return bytesRead;
}


-(void) sourceClose
{
    if (self.audioSource)
    {
       [self.audioSource sourceClose];

    }
}


// calculate the rms of a buffer containing 16 bit audio samples
-( float ) rms16: (signed char*) buffer  bufferSize:(NSUInteger) bufferSize
{
    float sum = 0.0;
    NSUInteger mySize = bufferSize / sizeof(short);
    short *buffer_short = (short*)buffer;
    
    for(NSUInteger i = 0; i < mySize; i++)
    {
       sum += (float)(buffer_short[i] * buffer_short[i]);
    }
    
    float sqrt1 = sqrtf( sum / mySize );
    
    return (float)(sqrt1)/mySize;
}

@end

File: gn-mobile-ios-1.2.1.4576/mmid/MusicIdSwift/GnAudioVisualizeAdapter.swift
//
//  GnAudioVisualizeAdapter.swift
//  GN_Music_SDK_iOS
//
//  Copyright (c) 2016 Gracenote. All rights reserved.
//

import CoreAudio
import AudioToolbox
//import GnAppDelegate

class GnAudioVisualizeAdapter : NSObject, GnAudioSourceDelegate
{
	var audioSource:GnAudioSourceDelegate!
	var numBitsPerSample:UInt!
	var numChannels:UInt!
	var lastRMS:Float!
	var maxRMS:Float!
	var audioVisualizerDelegate:GnAudioVisualizerDelegate!
	
	init( audioSource:GnAudioSourceDelegate,  audioVisualizerDelegate:GnAudioVisualizerDelegate )
	{
		self.lastRMS = 0
		self.audioSource = audioSource
		self.audioVisualizerDelegate = audioVisualizerDelegate
	}

	func sourceInit() -> UInt
	{
		if ( nil == audioSource )
		{
			return 1
		}
		
		// get format information for use later
		let retVal = audioSource.sourceInit()
		if ( retVal == 0 )
		{
			numBitsPerSample = audioSource.sampleSizeInBits()
			numChannels = audioSource.numberOfChannels()
		}
		
		return retVal;
	}

	func numberOfChannels() -> UInt
	{
		return numChannels
	}


	func sampleSizeInBits() -> UInt
	{
		return numBitsPerSample
	}

	func samplesPerSecond() -> UInt
	{
		if ( nil == audioSource )
		{
			return 0
		}
		
		return audioSource.samplesPerSecond()
	}

	func getData( dataBuffer: UnsafeMutablePointer<UInt8>, dataSize: UInt ) -> UInt
	{
		if ( nil == audioSource)
		{
			return 0;
		}
		
		let bytesRead = audioSource.getData( dataBuffer, dataSize:dataSize )
		let RMS = rms16( dataBuffer, dataSize:bytesRead )
		
//		if ( nil != audioVisualizerDelegate && audioVisualizerDelegate as? GnAudioVisualizerDelegate && audioVisualizerDelegate as? RMSDidUpdateByValue )
//if ( nil != audioVisualizerDelegate && nil != (audioVisualizerDelegate as? GnAudioVisualizerDelegate) )
//		{
//			audioVisualizerDelegate.RMSDidUpdateByValue(RMS/10)
//		}

		return bytesRead
	}

	func sourceClose()
	{
		if (nil == audioSource)
		{
		   audioSource.sourceClose()
		}
	}


	// calculate the rms of a buffer containing 16 bit audio samples
	func rms16( dataBuffer: UnsafeMutablePointer<UInt8>, dataSize: UInt ) -> Float
	{
//		var sum:Float = 0.0
//		var mySize:UInt = bufferSize / sizeof(short)
//		var bufferShort:UnsafeMutablePointer<UInt16> = dataBuffer
//		
//		for i in 0...mySize
//		{
//		   sum += (float)(bufferShort[i] * bufferShort[i]);
//		}
//		
//		let sqrt1 = sqrtf( sum / mySize );
//		return (Float)(sqrt1)/mySize;
		return 0
	}
}
